AISKU/src/Initialization.ts

```diff
@@ -365,17 +366,30 @@ export class Initialization implements IApplicationInsights {
                 }
             };
 
-            if (!Util.addEventHandler('beforeunload', performHousekeeping)) {
-                appInsightsInstance.appInsights.core.logger.throwInternal(
-                    LoggingSeverity.CRITICAL,
-                    _InternalMessageId.FailedToAddHandlerForOnBeforeUnload,
-                    'Could not add handler for beforeunload');
+            if (!appInsightsInstance.appInsights.config.disableFlushOnBeforeUnload) {
+                // Hook the unload event for the document, window and body to ensure that the client events are flushed to the server
+                // As just hooking the window does not always fire (on chrome) for page navigations.
+                let added = CoreUtils.addEventHandler('beforeunload', performHousekeeping);
+                added = CoreUtils.addEventHandler('pagehide', performHousekeeping) || added;
+                if (!added) {
```

Comment:
If pagehide fails and beforeunload succeeds would that be considered an error?

Reply:
No, we just need to attach to at least one of the events (i.e. whatever the browser supports) so that we ensure that the housekeeping is performed.

Reply:
Got it. So only 1 of these will be supported in each browser?

Reply:
Not exactly, some browsers do support both, but not all browsers.
e.g. 
pagehide == firefox and older IOS/MAC (< v12 - does not support beforeunload)
https://developer.mozilla.org/en-US/docs/Web/API/Window/pagehide_event

beforeunload == all modern and some older browser
https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event

Registering both is fine

Reply:
If both are registered then would the event handlers be called twice? It is fine as is since it is only calling flush currently, but if we were send an event in there in the future, would it be duped?

Reply:
Yes, this is something that we would need to deal with, but it's that case that we would always have to deal with it to ensure that we capture all events...
We *probably* should also be listening to the "unload" event -- just in case as well (we do in identity) but looking at the support matrix, all browsers that support unload also support beforeunload -- so it *should* be safe.